File: 04-eval/c/src/token.c
#include "token.h"
#include <string.h>

Token token_init(TokenType token_type, char* lit, int len ,int pos){
	Token token = {
		token_type,
		{0},
		len,
		pos,
	};
	if(token.len > MAX_LITERAL_LEN){
		token.len = MAX_LITERAL_LEN;
	}
	memcpy(&token.literal,lit,token.len);
	return token;
}

File: 04-eval/src/token.zig
pub const maxLiteralLen = 25;

pub const TokenType = enum{
	left_par,right_par,
	plus,star,slash,minus,
	double_star,double_slash,percent,
	number,
	illegal,too_long_number,
	eof,
};

pub const Token = struct{
    type: TokenType,
    literal: [maxLiteralLen:0]u8,
    pos: usize,
    len: usize,

    pub fn init(typ: TokenType, lit: []const u8, pos: usize) Token {
        var literal: [maxLiteralLen:0]u8 = undefined;
        const len = @min(lit.len, maxLiteralLen);
        @memcpy( literal[0..len], lit);
        if (len < maxLiteralLen) {
            @memset(literal[len..], 0);
        }
        return Token{
            .type = typ,
            .literal = literal,
            .pos = pos,
            .len = len,
        };
    }
};